src/Model/Scope/Condition.php

```diff
@@ -0,0 +1,384 @@
+            $this->operator = self::$opposites[$this->operator];
+        } else {
+            throw new Exception('Negation of condition is not supported for ' . ($this->operator ?: 'no') . ' operator');
+        }
+
+        return $this;
+    }
+
+    public function find($key): array
+    {
+        $field = $this->key;
+
+        if ($field instanceof Field) {
+            $field = $field->short_name;
+        } elseif (!is_string($field)) {
+            return [];
+        }
+
+        return ($field === $key) ? [$this] : [];
+    }
+
+    public function toWords(bool $asHtml = false): string
+    {
+        if (!$this->model) {
+            throw new Exception('Model must be set using setModel to convert to words');
+        }
+
+        // make sure clones are used to avoid changes
+        $condition = clone $this;
+
+        $key = $condition->keyToWords($asHtml);
+
+        $operator = $condition->operatorToWords($asHtml);
+
+        $value = $condition->valueToWords($condition->value, $asHtml);
+
+        $ret = trim("{$key} {$operator} {$value}");
+
+        return $asHtml ? $ret : html_entity_decode($ret);
+    }
+
+    protected function keyToWords(bool $asHtml = false): string
+    {
+        $model = $this->model;
+
+        $words = [];
+
+        if (is_string($field = $this->key)) {
+            if (stripos($field, '/') !== false) {
+                $references = explode('/', $field);
+
+                $words[] = $model->getModelCaption();
+
+                $field = array_pop($references);
+
+                foreach ($references as $link) {
+                    $words[] = "that has reference {$this->readableCaption($link)}";
+
+                    $model = $model->refLink($link);
+                }
+
+                $words[] = 'where';
+
+                if ($field === '#') {
+                    $words[] = $this->operator ? 'number of records' : 'any referenced record exists';
+                    $field = '';
+                } elseif ($field === '?') {
+                    $words[] = 'any referenced record exists';
+                    $field = '';
+                } elseif ($field === '!') {
+                    $words[] = 'no referenced records exist';
+                    $field = '';
+                }
+            }
+
+            $field = $model->hasField($field) ? $model->getField($field) : null;
+        }
+
+        if ($field instanceof Field) {
+            $words[] = $field->getCaption();
+        } elseif ($field instanceof Expression) {
+            $words[] = "expression '{$field->getDebugQuery()}'";
+        }
+
+        $string = implode(' ', array_filter($words));
+
+        return $asHtml ? "<strong>{$string}</strong>" : $string;
+    }
+
+    protected function operatorToWords(bool $asHtml = false): string
+    {
+        return $this->operator ? (self::$dictionary[strtoupper((string) $this->operator)] ?? 'is equal to') : '';
```

Comment:
hmm, operator should never be option

do we need to support unary expression?

if so, we should never add anything, maybe not even allow them to use in unary condition class design (more braninpower needed)

Reply:
Condition can be raw Expression. That is why iperator is not compulsory.